[WGSL] Add support for textures

tadeuzagallo opened the pull request:
#### ac45a24e64e7049b9fc1e19b7221d074f30f2d49
<pre>
[WGSL] Add support for textures
<a href="https://bugs.webkit.org/show_bug.cgi?id=254278">https://bugs.webkit.org/show_bug.cgi?id=254278</a>
&lt;rdar://problem/107067119&gt;

Reviewed by Mike Wyrzykowski.

Still needs the functions that use textures to make it useful, but this patch
adds support for textures in the parser, type checker and code generator.

* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::ParameterizedTypeName::stringViewToKind):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TextureKey::extra const):
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::textureType):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: <a href="https://commits.webkit.org/262063@main">https://commits.webkit.org/262063@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/222370d2d50d3353a414e735f73eb3093f1e0d0d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/367 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/343 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/420 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/417 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/584 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/371 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/356 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/353 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/396 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/343 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/412 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/323 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/360 "9 failures") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/347 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/336 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/315 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/354 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/105 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/352 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/AST/ASTTypeName.h:
```diff
@@ -103,9 +103,20 @@ class ParameterizedTypeName : public TypeName {
+        TextureStorage3d,
     };
-    static constexpr unsigned NumberOfBaseTypes = 12;
+    static constexpr unsigned NumberOfBaseTypes = 23;
```
instead of manually counting, should we just cast the last item, TextureStorage3d, to an `unsigned` int?

mwyrzykowski replied:
it may be just as problematic however since it is possible to forget to update it

djg replied:
@mwyrzykowski We can't cast `TestStorage3d` because it will be `22` not `23`. We could add `NumberOfTypes` (or equivalent WebKit approved naming) as the last element and cast `Base::NumberOfTypes` to `unsigned`?

mwyrzykowski replied:
If we add NumberOfTypes then we will end up causing compiler warnings / errors everywhere NumberOfTypes is not used, right? Which seems not worth it.

I am fine leaving it as-is, we could also cast the value and add 1 to it, but really it is fine as-is too

tadeuzagallo replied:
I considered that, but I don't know of any solution that doesn't require manual updates... I can change it to use the last value + 1. I think that's an improvement, since it doesn't require explicitly counting. I'd prefer that over adding the extra entry, since we'll need to handle it in every switch, as @mwyrzykowski mentioned.

tadeuzagallo replied:
Oh, we could also use macros, but I think not everyone is a fan 😬

djg reviewed (approved):
:shipit: with nit addressed.

tadeuzagallo committed:
[WGSL] Add support for textures
https://bugs.webkit.org/show_bug.cgi?id=254278
<rdar://problem/107067119>

Reviewed by Mike Wyrzykowski.

Still needs the functions that use textures to make it useful, but this patch
adds support for textures in the parser, type checker and code generator.

* Source/WebGPU/WGSL/AST/ASTTypeName.h:
(WGSL::AST::ParameterizedTypeName::stringViewToKind):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/TypeStore.cpp:
(WGSL::TextureKey::extra const):
(WGSL::TypeStore::TypeStore):
(WGSL::TypeStore::textureType):
* Source/WebGPU/WGSL/TypeStore.h:
* Source/WebGPU/WGSL/Types.cpp:
(WGSL::Type::dump const):
* Source/WebGPU/WGSL/Types.h:

Canonical link: https://commits.webkit.org/262063@main
